scripts/analyze_csv_data.py

```diff
@@ -0,0 +1,771 @@
+
+
+def plot_loghist(x, nbins):
+    x = x[~np.isnan(x)]
+    hist, bins = np.histogram(x, bins=nbins)
+    logbins = np.logspace(np.log10(bins[0] + 1e-6), np.log10(bins[-1]), nbins)
+    plt.hist(x, bins=logbins)
+    plt.xscale('log')
+
+
+def gen_ratio_histograms_split_tol(df, prefix, par):
+    tols = df['tolerance'].unique()
+    for t in tols:
+        gen_ratio_histograms(df[df['tolerance'] == t],
+                             prefix + f'_tol_{t:.0E}',
+                             'cumulative_kkt_matrix_passes',
+                             f'KKT matrix passes SGM{SGM_SHIFT}',
+                             KKT_PASSES_LIMIT,
+                             par)
+        gen_ratio_histograms(df[df['tolerance'] == t],
+                             prefix + f'_tol_{t:.0E}',
+                             'solve_time_sec',
+                             'Wall-clock time (secs)',
+                             TIME_LIMIT_SECS,
+                             par)
+
+
+def gen_ratio_histograms(df, prefix, xaxis, xlabel, limit, par):
+    assert len(df['experiment_label'].unique()) == 2
+
+    (l0, l1) = df['experiment_label'].unique()
+
+    def performance_ratio_fn(df, par):
+        df = df.reset_index()
+        assert len(df) <= 2
+
+        df0 = df[df['experiment_label'] == l0]
+        df1 = df[df['experiment_label'] == l1]
+
+        instance = df.instance_name.unique()
+
+        if len(df0) == 1 and df0['termination_reason'].iloc[0] == OPT:
+            kkt_passes_0 = df0[xaxis].iloc[0]
+        else:
+            kkt_passes_0 = par * limit
+
+        if len(df1) == 1 and df1['termination_reason'].iloc[0] == OPT:
+            kkt_passes_1 = df1[xaxis].iloc[0]
+        else:
+            kkt_passes_1 = par * limit
+
+        # if (df['termination_reason'] != OPT).any():
+        #    return np.nan
+        return (kkt_passes_0 / kkt_passes_1)
+
+    ratios = df.groupby(['instance_name']) \
+        .apply(lambda _: performance_ratio_fn(_, par)) \
+        .reset_index(name='ratio')
+    plt.figure(figsize=(10, 6))
+    plt.title(sanitize_title(
+        f'{prefix} {xlabel} {label_lookup(l0)}:{label_lookup(l1)}'))
+    plot_loghist(ratios['ratio'], min(len(ratios) // 3, 25))
+    path = os.path.join(
+        FIGS_DIR,
+        f'{prefix}_{l0}_{l1}_{xaxis}_performance_ratio.pdf')
+    plt.savefig(path)
+    table = ratios.to_latex(float_format="%.2f",
+                            longtable=False,
+                            index=False,
+                            caption=f'Performance ratio.',
+                            label=f't:ratio-{prefix}',
+                            column_format='lc',
+                            na_rep='-')
+    table = change_table_font_size(table)
+    path = os.path.join(TEX_DIR, f'{prefix}_{l0}:'
+                                 f'{l1}_{xaxis}_ratio_table.tex')
+    with open(path, "w") as f:
+        f.write(table)
+
+# Unsolved problems might be missing from csv, make sure all are accounted for
```

Comment:
This comment seems like it applies to fill_in_missing_problems().  In that case, the two blank lines between the comment and the function should be deleted.